policy/engine.go

```diff
@@ -0,0 +1,838 @@
+	return true, nil
+}
+
+// IsIPAllowed verifies a single IP domain is allowed.
+func (e *NamePolicyEngine) IsIPAllowed(ip net.IP) (bool, error) {
+	if err := e.validateNames([]string{}, []net.IP{ip}, []string{}, []*url.URL{}, []string{}); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// ArePrincipalsAllowed verifies that all principals in an SSH certificate are allowed.
+func (e *NamePolicyEngine) ArePrincipalsAllowed(cert *ssh.Certificate) (bool, error) {
+	dnsNames, ips, emails, principals, err := splitSSHPrincipals(cert)
+	if err != nil {
+		return false, err
+	}
+	if err := e.validateNames(dnsNames, ips, emails, []*url.URL{}, principals); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// appendSubjectCommonName appends the Subject Common Name to the appropriate slice of names. The logic is
+// similar as x509util.SplitSANs: if the subject can be parsed as an IP, it's added to the ips. If it can
+// be parsed as an URL, it is added to the URIs. If it contains an @, it is added to emails. When it's none
+// of these, it's added to the DNS names.
+func appendSubjectCommonName(subject pkix.Name, dnsNames *[]string, ips *[]net.IP, emails *[]string, uris *[]*url.URL) {
+	commonName := subject.CommonName
+	if commonName == "" {
+		return
+	}
+	subjectDNSNames, subjectIPs, subjectEmails, subjectURIs := x509util.SplitSANs([]string{commonName})
+	*dnsNames = append(*dnsNames, subjectDNSNames...)
+	*ips = append(*ips, subjectIPs...)
+	*emails = append(*emails, subjectEmails...)
+	*uris = append(*uris, subjectURIs...)
+}
+
+// splitPrincipals splits SSH certificate principals into DNS names, emails and usernames.
+func splitSSHPrincipals(cert *ssh.Certificate) (dnsNames []string, ips []net.IP, emails, principals []string, err error) {
+	dnsNames = []string{}
+	ips = []net.IP{}
+	emails = []string{}
+	principals = []string{}
+	var uris []*url.URL
+	switch cert.CertType {
+	case ssh.HostCert:
+		dnsNames, ips, emails, uris = x509util.SplitSANs(cert.ValidPrincipals)
+		if len(uris) > 0 {
+			err = fmt.Errorf("URL principals %v not expected in SSH host certificate ", uris)
+		}
+	case ssh.UserCert:
+		// re-using SplitSANs results in anything that can't be parsed as an IP, URI or email
+		// to be considered a username principal. This allows usernames like h.slatman to be present
+		// in the SSH certificate. We're exluding IPs and URIs, because they can be confusing
+		// when used in a SSH user certificate.
+		principals, ips, emails, uris = x509util.SplitSANs(cert.ValidPrincipals)
+		if len(uris) > 0 {
+			err = fmt.Errorf("URL principals %v not expected in SSH user certificate ", uris)
+		}
+	default:
+		err = fmt.Errorf("unexpected SSH certificate type %d", cert.CertType)
+	}
+
+	return
+}
+
+// validateNames verifies that all names are allowed.
+// Its logic follows that of (a large part of) the (c *Certificate) isValid() function
+// in https://cs.opensource.google/go/go/+/refs/tags/go1.17.5:src/crypto/x509/verify.go
```

Comment:
Move all these methods to a different file and add 
```
// Copyright 2011 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
```

Reply:
👍